[nomerge] Oss517 - for diff only

ankona opened the pull request.

ankona committed:
initial telemetry monitor commit

ankona committed:
start tracking & using task status, start using on_timestep

ankona committed:
refactor dicts to dataclasses

ankona committed:
add launcher manual creation, start to see if i can swap to job manager

ankona committed:
checkpoint, wlm happy path only

remove demo exp creation

ankona committed:
ensure manifest is processed on load

ankona committed:
remove commented lines

ankona committed:
move file-write to job start callback

ankona committed:
add timestep hook, remove commented code

ankona committed:
pass injected logger to tracking methods

ankona committed:
Add tracking metadata to steps

ankona committed:
use pathlike hint instead of int to enable passing path

ankona committed:
add local step indirect execution

ankona committed:
remove infinite recursion bug

ankona committed:
minor cleanup

ankona committed:
minor cleanup & commenting

ankona committed:
minor organization, add indirect.py tests

ankona committed:
add watchdog to installed dependencies

ankona committed:
add tests, minor fixes, add loop limit to monitor for testing

ankona committed:
comments

ankona committed:
lint/mypy

ankona committed:
fix for mypy/watchdog

ankona committed:
telemetry monitor fixes. untested

ankona committed:
test fixes

ankona committed:
fix is-managed computed prop

ankona committed:
all telem mon tests working

ankona committed:
fix incorrect proxy-module name after move

ankona committed:
fix bad outputs from indirect

ankona committed:
fix missing test args for indirect

ankona committed:
fix tests not b64 encode cmd

ankona committed:
fix naming bug on colo launch script, fix indirect launch bug

ankona committed:
tweak test assertion

ankona committed:
avoid overwriting model outputs

ankona committed:
fix file overwrite bug, remove commented code block

ankona committed:
fix lint issue, fix missing timestamp in tracking file

ankona committed:
use flag to optionally execute telemetry proxy

ankona committed:
formatting

ankona committed:
remove extraneous wait, ensure cleanup

ankona committed:
post-rebase fix

ankona committed:
lint/merge fixes

ankona committed:
additional lint/mypy/merge fixes

ankona committed:
add bug fix from mypy tweak

ankona committed:
avoid unnecessary pylint mitigation

ankona committed:
shift ctx managers

ankona committed:
relocate some helper methods, minor cleanup

ankona committed:
handle iterable entities

ankona committed:
fix bug attempting to stop inactive fs listener

ankona committed:
ignore output files from debug sessions

ankona committed:
fix job/step id logging bug

ankona committed:
fix incorrect import bug

ankona committed:
lint & typing

ankona committed:
update for new metadata format

ankona committed:
autostart

ankona committed:
fix global var access bug

ankona committed:
remove unused process functions

ankona committed:
remove call to commented code

ankona committed:
static analysis fix

ankona committed:
auto start/stop telemetry manager

ankona committed:
fix rebase bugs

ankona committed:
create telmon callback w/experiment directory

ankona committed:
undo db entrypoint rename

ankona committed:
remove commented code blocks

ankona commented on smartsim/_core/launcher/local/local.py:
```diff
@@ -104,16 +115,23 @@ def run(self, step: Step) -> str:
         passed_env = step.env if isinstance(step, LocalStep) else None
 
+        cmd = step.get_launch_cmd()
+        if os.environ.get("SSFLAG_TELEMETRY", False):
```
@MattToast suggestions on env var?

MattToast replied:
I think the name is great! It leaves it open for other additional `SSFLAG_*` vars in future which I like!

One small note: if  a user runs `export SSFLAG_TELEMETRY=0` this condition will evaluate true bc `"0"` is a non-empty string. Should we update this so that it first casts the env to an int to handle that case?

ankona replied:
per later convo, renamed.

ankona committed:
remove values no longer in manifest

ankona committed:
avoid raising unnecessary exceptions

ankona committed:
avoid raising unnecessary exception

ankona committed:
use constants to avoid redefining path to manifest

ankona committed:
fix bug where doubled events can overwrite existing files

MattToast commented on smartsim/_core/utils/helpers.py:
```diff
@@ -73,7 +75,7 @@ def create_lockfile_name() -> str:
 def check_dev_log_level() -> bool:
     try:
-        lvl = os.environ["SMARTSIM_LOG_LEVEL"]
+        lvl = os.environ.get("SMARTSIM_LOG_LEVEL", "")
```
key error can no longer be thrown. Feel free to remove the try/except

ankona replied:
Done

MattToast commented on smartsim/_core/control/controller.py:
```diff
@@ -74,6 +87,43 @@
 
 # job manager lock
 JM_LOCK = threading.RLock()
+TELEMETRY_MONITOR: t.Optional[t.Any] = None
```
Can we type hint this as `Optional[subprocess.Popen]`?

MattToast replied:
Also, incredibly pedantic change request, since nobody should be importing the telemetry monitor out of this module can we prefix its name with `_`? We should have done that with `JM_LOCK` also.

ankona replied:
renamed but need to still solve the generic stub issue

ankona replied:
solved stub issue w/if t.TYPE_CHECKING

MattToast commented on smartsim/_core/entrypoints/telemetrymonitor.py:
```diff
@@ -0,0 +1,654 @@
+from smartsim._core.control.job import Job, JobEntity
+from smartsim._core.control.jobmanager import JobManager
+from smartsim._core.launcher.stepInfo import StepInfo
+from smartsim._core.launcher.cobalt.cobaltLauncher import CobaltLauncher
```
super knit-picky style thing, but is there any reason the cobalt launcher is imported here and not in the section with all of the launchers?

ankona replied:
love it. moved.

MattToast commented on smartsim/_core/control/job.py:
```diff
@@ -27,10 +27,43 @@
+
+class JobEntity:
+    """
+    Minimum entity API required for Job to do what it does...
```
LMAO I love the description and pains me to ask this: but any chance I could twist your arm to make this a hair more descriptive?

ankona replied:
added detail on use as telemetry monitor data: 
```
"""Minimum API required for a job processed in the JobManager with support for
    telemetry monitoring"""
```

ankona committed:
remove iteration limiter used for testing purposes

MattToast commented on smartsim/_core/entrypoints/telemetrymonitor.py:
```diff
@@ -0,0 +1,654 @@
+    """Map run data persisted in a manifest file to an object"""
+    runs = [
+        Run(
+            timestamp=instance["run_id"],
```
Just a heads up: The timestamp got promoted to its own field in the `manifest.json` file. the key should just be `"timestamp"`

ankona replied:
tys. swapped to timestamp

ankona replied:
:+1:

MattToast commented on smartsim/_core/entrypoints/telemetrymonitor.py:
```diff
@@ -0,0 +1,654 @@
+            entity.name = persistable_entity["name"]
+            entity.job_id = str(metadata.get("job_id", ""))
+            entity.step_id = str(metadata.get("step_id", ""))
+            entity.timestamp = int(persistable_entity.get("run_id", "0"))
```
ditto here, `run_id` is now "uuid" (first 7 chars of a uuid on my branch, I think the full thing here), and no guarantee that they are all `int`s. Should be safe to replace this with `"timestamp"`

MattToast replied:
Actually, my understanding is that `persistable_entity` is a `dict` representing either an `Orchestrator` or an `Ensemble` in the `manifest.json`. Currently there is no `"run_id"` or `"timestamp"` field on these dicts, so this (currently) should always be defaulting to `"0"`.

ankona committed:
Remove startup hack; use experiment dir as base monitor path

ankona committed:
add missing docstrings, black formatting

MattToast commented on smartsim/_core/entrypoints/telemetrymonitor.py:
```diff
@@ -0,0 +1,654 @@
+    """Map entity data persisted in a manifest file to an object"""
+    entities: t.List[JobEntity] = []
+
+    if "shards" in persistable_entity or "models" in persistable_entity:
+        container = "shards" if "shards" in persistable_entity else "models"
```
Not a you problem, but would it be helpful to unify these keys into a single `"elements"` key to avoid this little roundabout way of accessing the correct field? Maybe worth a new ticket?

ankona replied:
used set intersection to clarify the key lookup:

```
    db_keys = {"shards", "models"}
    db_keys = db_keys.intersection(persistable_entity.keys())
    if db_keys:
```

added `_hydrate_persistable` to avoid the duplication:
```
    db_keys = {"shards", "models"}
    db_keys = db_keys.intersection(persistable_entity.keys())
    if db_keys:
        container = "shards" if "shards" in db_keys else "models"
        for shard in persistable_entity[container]:            
            entity = _hydrate_persistable(shard, entity_type, str(exp_dir))
            entities.append(entity)

        return entities

    entity = _hydrate_persistable(persistable_entity, entity_type, str(exp_dir))
    entities.append(entity)
```

ankona replied:
I think using the set should clean it up... Since the events don't really use anything other than the Job metadata, I'd like to avoid adding any more complexity to the abstraction as long as possible.

MattToast commented on smartsim/_core/entrypoints/telemetrymonitor.py:
```diff
@@ -0,0 +1,676 @@
+            entity.type = entity_type
+            entity.name = persistable_entity["name"]
+            entity.job_id = str(metadata.get("job_id", ""))
+            entity.step_id = str(metadata.get("step_id", ""))
```
I may have accidentally changed these field names out from underneath you 😬. I am currently writing out `"step_id"` and `"task_id"` to more closely follow the `StepMap` class.

MattToast replied:
Also these keys will always be there, but they are (currently) nullable, so you could end up with some `JobEntity` where
```py
assert hydrate_persistable("model", instance, exp_dir).pop().job_id == "None"
```

ankona replied:
I'm inspecting the parents for ensembles & orchestrators and creating a persistable for the child entity. I updated the set var name to more accurately reflect that. See body of `hydrate_persistables`:

```
# an entity w/parent key creates persistables for entities it contains
parent_keys = {"shards", "models"}
parent_keys = parent_keys.intersection(persistable_entity.keys())
```

ankona replied:
- I updated the keys being read but didn't update property names - the `Job` class uses `job_id` in place of `task_id`. For now, I'd rather keep the two matching just to avoid more confusion.
- Added some cleanup to replace "None" w/empty string

MattToast commented on smartsim/_core/entrypoints/telemetrymonitor.py:
```diff
@@ -0,0 +1,676 @@
+    entity.step_id = str(metadata.get("step_id", ""))
+    entity.timestamp = int(persistable_entity.get("run_id", "0"))
+    entity.path = str(exp_dir)
+    entity.status_dir = str(status_dir)
```
This block here looks near identical to lines 116-125. Any chance we could extract some of this logic into a helper fn or something?

ankona replied:
I saw the same thing after your earlier comment! Good eye. I added `_hydrate_persistable` for this

MattToast commented on smartsim/_core/launcher/local/local.py:
```diff
@@ -127,11 +145,50 @@ def stop(self, step_name: str) -> UnmanagedStepInfo:
         return "Local"
+
+    @classmethod
+    def get_proxy_cmd(cls, step: Step) -> t.List[str]:
```
Not particularly important, but `cls` is unused. You should be free to make this a static method if you'd like

ankona replied:
+1. Done!

ankona committed:
remove unused type ignore

ankona committed:
linter fix (too many statements in try)

MattToast commented on smartsim/_core/control/job.py:
```diff
@@ -27,10 +27,43 @@
 from ...status import STATUS_NEW
 
 
+_JobKey = t.Tuple[str, str]
```
Since we are assigning this a custom type alias anyway, would there be interest in promoting this to a `NamedTuple` just so that we can access the fields of the tuple with `.` notation?

ankona replied:
_love_ it.

MattToast commented on smartsim/_core/control/jobmanager.py:
```diff
@@ -76,12 +81,18 @@ def __init__(self, lock: RLock, launcher: t.Optional[Launcher] = None) -> None:
         self._lock = lock  # thread lock
 
         self.kill_on_interrupt = True  # flag for killing jobs on SIGINT
+        self.on_complete_hook: t.List[t.Callable[[Job, Logger], None]] = []
+        self.on_start_hook: t.List[t.Callable[[Job, Logger], None]] = []
+        self.on_timestep_hook: t.List[t.Callable[[Job, Logger], None]] = []
 
-    def start(self) -> None:
+    def start_job_monitor(self) -> None:
         """Start a thread for the job manager"""
         self.monitor = Thread(name="JobManager", daemon=True, target=self.run)
         self.monitor.start()
 
+    def start(self) -> None:
+        self.start_job_monitor()
```
Any reason for this little refactor? From what I can tell `start_job_monitor` is only ever called in `start`

ankona replied:
Good question. I recall wanting to start the job monitor independently but i think that fell apart later. Reverting change.

MattToast commented on smartsim/entity/dbnode.py:
```diff
@@ -76,8 +76,12 @@ def __init__(
 
     @property
     def num_shards(self) -> int:
+        if not hasattr(self.run_settings, "mpmd"):
```
Now that this check is here, you should be safe to remove the `try/except` here. Ditto for `is_mpmd`

ankona replied:
removed & removed

MattToast commented on smartsim/_core/utils/helpers.py:
```diff
@@ -273,3 +275,30 @@ def installed_redisai_backends(
+
+
+def get_ts() -> int:
+    """Helper function to ensure all timestamps are converted to integers"""
```
could you just throw a unit in this docstr rq. I wasn't immediately confident if this evaluated to seconds, nanoseconds, or some other unit.

ankona replied:
100% - I think we had a prior convo about needing ms accuracy so I might need to change this anyway?

MattToast commented on smartsim/_core/utils/helpers.py:
```diff
@@ -273,3 +275,30 @@ def installed_redisai_backends(
+    return int(datetime.timestamp(datetime.now()))
+
+
+def encode_cmd(cmd: t.List[str]) -> str:
```
@MattToast probably a good idea to write up a ticket to use these fns for our other entry points that take lists of strs (`redis.py`/`colocated.py`), but definitely out of scope for this PR.

ankona replied:
I encoded those CLI params to avoid a headache of having some unknown "user program" w/a weird argument cause some strange problem passing arguments through. With redis/colocated, we know what those arguments formats are and shouldn't see the same type of issue. 

I don't think we'll need to reuse this (and per our offline convo, we may choose to switch to "-- <extra args>" to avoid encoding altogether).

ankona committed:
ensure feature flag behaves as expected when disabled...

...e.g. os.environ["SSFLAG_TELEMETRY"] == "0"

MattToast commented on smartsim/_core/control/jobmanager.py:
```diff
@@ -135,6 +148,9 @@ def move_to_completed(self, job: Job) -> None:
             elif job.ename in self.jobs:
                 del self.jobs[job.ename]
 
+            # if not self.jobs and not self.db_jobs:
+            #     self.stop_telemetry_monitor()
```
looks like old debug code that can be removed

ankona replied:
good :eye:

MattToast commented on smartsim/_core/control/controller.py:
```diff
@@ -74,6 +87,43 @@
+        ],
+        stdin=subprocess.PIPE,
+        stdout=subprocess.PIPE,
+        cwd=str(pathlib.Path(__file__).parent.parent.parent),
```
Not wrong just curious: any reason this needs to be run from `smartsim` dir over the cwd?

MattToast commented on smartsim/_core/control/controller.py:
```diff
@@ -250,6 +304,19 @@ def get_jobs(self) -> t.Dict[str, Job]:
+    def get_db_jobs(self) -> t.Dict[str, t.Tuple[Job, t.Union[DBNode, Orchestrator]]]:
+        """Return a dictionary of database job data
+
+        :returns: dict[str, Job]
```
mismatched return typehint vs `:returns:` docstr

MattToast replied:
@MattToast remember to look into a sphinx plugin to automatically generate signature for docstrs.

ankona replied:
fixed & +1 on autodoc

MattToast commented on smartsim/_core/control/controller.py:
```diff
@@ -436,7 +503,7 @@ def _launch_orchestrator(
 
         # set the jobs in the job manager to provide SSDB variable to entities
         # if _host isnt set within each
-        self._jobs.set_db_hosts(orchestrator)
+        self._set_db_hosts(orchestrator)
```
Now that we have (in my opinion correctly) moved the logic to set db hosts from the JM to the controller, you should be free to remove the `JobManager.set_db_hosts` method, as from what I can tell it is never called in SS

ankona replied:
Done!

MattToast commented on smartsim/_core/control/controller.py:
```diff
@@ -727,6 +796,40 @@ def reload_saved_db(self, checkpoint_file: str) -> Orchestrator:
 
             return orc
 
+    def _get_db_host_addresses(self) -> t.List[str]:
```
Looks like`JobManager.get_db_host_addresses` is still being used at lines 407, 619, 646 and 837. It also looks like this implementation is never called in SS. Feel free to either rm this fn or update the mentioned lines and remove the original implementation to remove the duplication!

ankona replied:
merge fail. fixed by removing original that was re-introduced

ankona committed:
remove deprecated test

MattToast commented on smartsim/_core/entrypoints/telemetrymonitor.py:
```diff
@@ -0,0 +1,676 @@
+    Persist a tracking event for an entity
+    """
+    job_id = job_id or ""
+    step_id = step_id or ""
```
redundant according to type hints. The only way a false-y value will get passed is is if it is already an empty str

ankona replied:
removed

MattToast commented on smartsim/_core/control/jobmanager.py:
```diff
@@ -24,21 +24,26 @@
-from .job import Job
+from .job import Job, JobEntity
+
+# from ..entrypoints.telemetrymonitor import start_monitor, kill_monitor
```
still need?

ankona replied:
removed!

MattToast commented on smartsim/_core/control/jobmanager.py:
```diff
@@ -166,7 +177,7 @@ def add_job(
         job_name: str,
         job_id: t.Optional[str],
-        entity: t.Union[SmartSimEntity, EntitySequence[SmartSimEntity]],
+        entity: t.Union[SmartSimEntity, EntitySequence[SmartSimEntity], JobEntity],
```
Definitely out of scope for this PR, but would it make sense to make a typing protocol specifically for bare minimum set of attrs that the JM needs to track an entity? Just to avoid the super long unions?

ankona replied:
Due to existence of `get_db_host_addresses` & `set_db_hosts` on `JobManager`, I was unable to make that swap. If we move those methods to `Controller`, the `JobManager` really doesn't have anything holding this back. I'll create a ticket.


** I had previously moved the methods but reverted due to testing issues. This ticket will give a nicer place to repeat the move

ankona replied:
Ticket 569 created.

AlyssaCote commented on tests/test_configs/manifest_enh.json:
```diff
@@ -0,0 +1,63 @@
+    "runs": [
+        {
+            "applications" : [
+                {
```
is this manifest supposed to be malformed?

ankona replied:
it could be out of date... offline sync incoming...

ankona replied:
unused mock data removed

MattToast commented on smartsim/_core/entrypoints/telemetrymonitor.py:
```diff
@@ -0,0 +1,676 @@
+        :raises TypeError: if no launcher argument is provided.
+        """
+
+        if launcher is not None:
```
pedantic typing thing: clause is theoretically unfalsifiable according to type hint

ankona replied:
copypasta from jobmanager.py -> removed from telemetrymonitor.py

MattToast commented on smartsim/_core/entrypoints/telemetrymonitor.py:
```diff
@@ -0,0 +1,676 @@
+            self._logger.error("Manifest content error", exc_info=True)
+            return
+
+        self.set_launcher(manifest.launcher)
```
At time of writing, in the manifest, this string may have capitals as it is gathered by casting an instance of the launcher class to a str (for some reason all of the launchers have `__str__` overwritten). Might be worthwhile to specify `self.set_launcher(manifest.launcher.lower())`

ankona replied:
the launcher instantiation code already handles calling that .lower() - see: `init_launcher`

![image](https://github.com/ankona/SmartSim/assets/3595025/c58f6471-8472-4c3a-85e3-d7e7c84b9d6e)

MattToast commented on smartsim/_core/entrypoints/telemetrymonitor.py:
```diff
@@ -0,0 +1,676 @@
+
+        self.set_launcher(manifest.launcher)
+
+        if not self.job_manager._launcher:  # pylint: disable=protected-access
```
If you want to mitigate the pylint error, presumably this is the same launcher as `self._launcher` right?

ankona replied:
Great catch. Replaced private usage with `self._launcher`

MattToast commented on smartsim/_core/entrypoints/telemetrymonitor.py:
```diff
@@ -0,0 +1,676 @@
+            experiment_dir,
+            self._logger,
+            detail=detail,
+        )
```
Will this `track_event` call do anything if the `track_completed` callback is called in `self.job_manager.move_to_completed`? It looks like it will do some formatting but wont dump anything to disk because the hook will have placed a file with the same name and `track_event` fn will silently NOT overwrite that file.

ankona replied:
Removed registration of the onstart and onstop callback in the telemetry monitor

MattToast commented on smartsim/_core/entrypoints/telemetrymonitor.py:
```diff
@@ -0,0 +1,676 @@
+                    entity.job_id,
+                    entity,
+                    entity.is_managed,
+                )
```
Similar to the previous comment, will the "on created" hook in the JM do anything if the manual call to `track_event` here creates a file in the same spot in with the same name that the hook is unwilling to overwrite?

ankona replied:
removed registration of hook registrations.

MattToast commented on smartsim/_core/control/controller.py:
```diff
@@ -110,6 +160,10 @@ def start(
         signal.signal(signal.SIGINT, self._jobs.signal_interrupt)
+
+        # have job manager launch a telemetry monitor if any jobs are run
+        self._jobs.add_job_onstart_callback(create_telmon_callback(exp_path))
```
Not necessarily wrong, but I thought I should call it out: if there are multiple calls to `start`, because `create_telmon_callback` creates a new fn each time it is called, this will cause the `self._jobs.on_start_hook` to grow indefinitely (or rather 1 entry per 1 `start` call).

ankona replied:
Great catch. the controller is used to automatically start the telemetry monitor after it shuts itself down and doesn't have any other onstart callbacks for the job manager.

It may not work if we add more callbacks, but for now I will ensure it isn't duped by checking if the controllers job manager already has an onstart callback.

MattToast commented on smartsim/_core/launcher/local/local.py:
```diff
@@ -104,16 +115,24 @@ def run(self, step: Step) -> str:
         passed_env = step.env if isinstance(step, LocalStep) else None
 
+        cmd = step.get_launch_cmd()
+        is_enabled = int(os.environ.get("SSFLAG_TELEMETRY", 0))
```
I know I signed off on this name before, but after thinking it over a bit more, can we actually change it to `SMARTSIM_FLAG_TELEMETRY` just to make it hair more consistent with the existing `SMARTSIM_*` env vars.

I also like the idea of establishing the pattern of `SMARTSIM_*` env vars are ones that we expect users to touch and `SS*` vars like `SSDB` are ones that users should not set.

MattToast commented on smartsim/_core/entrypoints/telemetrymonitor.py:
```diff
@@ -0,0 +1,676 @@
+            self.job_manager.set_launcher(self._launcher)
+            self.job_manager.add_job_onstart_callback(track_started)
+            self.job_manager.add_job_onstop_callback(track_completed)
+            self.job_manager.add_job_onstep_callback(track_timestep)
```
I'm not sure if I have something set up in my env wrong, but for the life of me I cannot seem to get the `track_timestep` callback to fire. Can you try running this quick experiment and see if anything is produced?
```py
from smartsim import Experiment
exp = Experiment("MyExperiment", launcher="slurm")

sleep_rs = exp.create_run_settings("sleep", ["30"])
sleep_rs.set_nodes(1)
sleep_rs.set_tasks(1)

sleep = exp.create_model("SleepModel", sleep_rs)
exp.start(sleep, block=True)
```
If it is working as expected on your end, it might be worth syncing over this real quick.

ankona replied:
there was an issue i had to fix yesterday w/ the watch directory & the watch pattern (if dir didn't exist, the watch configuration would fail). swapping to the exp_dir and using the relative path to the manifest in the pattern fixed this. let's retest together to be sure you see what i do

MattToast commented on smartsim/_core/entrypoints/indirect.py:
```diff
@@ -0,0 +1,213 @@
+                str(STEP_PID),
+                etype,
+                "stop",
+                exp_path,
```
It looks like these event files are being created, but are ending up outside of the top level `{experiment_dir}/.smartsim` directory. For instance, in this experiment:
```py
from smartsim import Experiment

exp = Experiment("MyExperiment", launcher="local")

sleep_rs = exp.create_run_settings("sleep", ["30"])
sleep = exp.create_model("SleepModel", sleep_rs)
exp.generate(sleep)

exp.start(sleep, block=True)
```
it looks like the following files were created
```
MyExperiment/SleepModel/.smartsim/telemetry/model/SleepModel-CWGDWZOUWOBH/start.json
MyExperiment/SleepModel/.smartsim/telemetry/model/SleepModel-CWGDWZOUWOBH/stop.json
MyExperiment/.smartsim/telemetry/manifest.json
MyExperiment/.smartsim/telemetry/model/SleepModel/start.json
MyExperiment/.smartsim/telemetry/model/SleepModel/stop.json
```
do we need both of the `start/stop.json` files? If so, is there anyway we can bring them all under the `.smartsim` dir?

ankona replied:
paths updated to use `status_dir` instead of old compute w/`exp_dir`

MattToast commented on smartsim/_core/entrypoints/telemetrymonitor.py:
```diff
@@ -0,0 +1,676 @@
+    entity_type = etype or "missing_entity_type"
+
+    name: str = ename or "entity-name-not-found"
+    tgt_path = exp_dir / TELMON_SUBDIR / entity_type / name / f"{action}.json"
```
This comment may be moot considering my previous question in the last comment, but should we add directory in target path for `run_id` so that we do not run into name conflicts if someone were to restart a run? for e.g.
```py
from smartsim import Experiment

exp = Experiment("MyExperiment", launcher="local")

sleep_rs = exp.create_run_settings("sleep", ["30"])
sleep = exp.create_model("SleepModel", sleep_rs)
exp.generate(sleep)

exp.start(sleep, block=True)
exp.start(sleep, block=True)
# ^^^ Despite being started twice
# this will only produce one start and stop json file
```

ankona replied:
ditto on updated paths for use of `status_dir` instead of computing w/`exp_dir`

MattToast commented on smartsim/_core/entrypoints/telemetrymonitor.py:
```diff
@@ -0,0 +1,676 @@
+
+        names = {entity.name: entity for entity in entity_map.values()}
+
+        if launcher and names:
```
No need to check against launcher here, it will always come back non-nullable

ankona replied:
removed extraneous conditional

MattToast reviewed (changes requested):
This PR was awesome to play around with! Everything seems to be dumping files and properly reporting information when launching entities both against the local and WLM launchers!

I left a few comments for small little clean ups and some points where I think I may have seen multiple solutions trying to solve the same problem. I also left a couple questions regarding architecture just to make sure I have the design straight in my own head. Otherwise this seems to be working on my end!

ankona committed:
remove unnecessary try/catch

ankona committed:
rename var

ankona committed:
ensure logical import grouping

ankona committed:
add docstring detail

ankona committed:
use timestamp where appropriate in place of run_id

ankona committed:
avoid duplicate entity hydration code

ankona committed:
swap class- to staticmethod where class unused

ankona committed:
revert unnecessary method creation

AlyssaCote commented on smartsim/_core/entrypoints/telemetrymonitor.py:
```diff
@@ -0,0 +1,677 @@
+    entity_type = etype or "missing_entity_type"
+
+    name: str = ename or "entity-name-not-found"
+    tgt_path = exp_dir / TELMON_SUBDIR / entity_type / name / f"{action}.json"
```
Where are shard statuses currently being written to? I thought the path was something like:
`exp_dir/.smartsim/telemetry/<run_id>/<entity_type>/<orc_name> /<shard_id>/start.json `

ankona replied:
leftover from target dir swap. fixed to write to `entity.status_dir` instead

ankona committed:
remove deprecated try/catch

ankona committed:
Convert jobkey to named tuple

ankona committed:
remove commented code block

ankona committed:
add units to docstring for clarity

ankona committed:
fix docstring type mismatch

ankona committed:
rename feature flag

ankona committed:
remove always true conditional

ankona committed:
remove relocated methods

ankona committed:
fixes: use status_dir as write target, clear "None" task/step id

ankona committed:
fix output path bug

ankona committed:
fix data format change

ankona committed:
descriptive commenting, general var naming

ankona committed:
remove unused param & update tests

ankona committed:
remove unused params, fix write paths

ankona committed:
remove unnecessary conditional

ankona committed:
more key drift, fix computed test output path

ankona committed:
path drift fixes

ankona committed:
remove unused test config

ankona committed:
use local member instead of private launcher

ankona committed:
remove unnecessary data clean

ankona committed:
use local instead of private ref

ankona committed:
remove on_start, on_stop callback registrations (duped in telmon)

ankona committed:
popen type workaround

ankona committed:
fix looping value bug

ankona committed:
undo method move

ankona committed:
avoid duplicated effort to track job events

ankona committed:
ensure multiple telemetry managers are not started

ankona committed:
Fix text concat in log msg

ankona committed:
unmanaged proxy path hack

ankona committed:
indirect process tracking hack

ankona committed:
formatting

ankona committed:
fix incorrect job_id passed

ankona committed:
bugfix (result is not none)

ankona committed:
enable non-uniform subitem parsing

ankona committed:
fix bug where dbnode lookups fail

ankona committed:
fix incorrect pipe

ankona committed:
fix defaults, remove commented code

ankona committed:
remove unused import

ankona committed:
update manifest test data & tests accordingly

ankona committed:
reverse step & task id when deserializing

ankona committed:
don't monitor unmanaged tasks w/telmon service

ankona committed:
fix failure to auto-shutdown

ankona committed:
sanity check job id eval

ankona committed:
enable retry manifest loading

ankona committed:
log warning when event writes fail

ankona committed:
add default output path for indirect proxy

ankona committed:
remove extraneous var

ankona committed:
formatting

ankona committed:
pass status_dir to indirect as CLI arg

ankona committed:
fix incorrect is_task value for managed job

ankona committed:
lint/mypy

ankona committed:
ensure telemetry proxy doesn't stomp expected outputs

ankona committed:
fix fileno bug

MattToast committed:
Add status dir to meta attr for step

AlyssaCote commented on smartsim/_core/entrypoints/telemetrymonitor.py:
```diff
@@ -0,0 +1,710 @@
+        logger.error("Unable to write tracking file.", exc_info=True)
+
+
+def track_completed(job: Job, logger: logging.Logger) -> None:
+    """Persists telemetry event for the end of job"""
+    detail = job.status
+
+    if hasattr(job.entity, "status_dir"):
+        write_path = pathlib.Path(job.entity.status_dir)
+    else:
+        write_path = pathlib.Path(job.entity.path)
+
+    track_event(
+        get_ts(),
+        job.jid or "" if not job.is_task else "",
+        job.jid or "" if job.is_task else "",
+        job.entity.type,
+        "stop",
+        write_path,
+        logger,
+        detail=detail,
+    )
+
+
+def track_started(job: Job, logger: logging.Logger) -> None:
+    """Persists telemetry event for the start of job"""
+    if hasattr(job.entity, "status_dir"):
+        write_path = pathlib.Path(job.entity.status_dir)
+    else:
+        write_path = pathlib.Path(job.entity.path)
+
+    track_event(
+        get_ts(),
+        job.jid or "" if not job.is_task else "",
+        job.jid or "" if job.is_task else "",
+        job.entity.type,
+        "start",
+        write_path,
+        logger,
+    )
```
These are defined but never used. Are they still needed?

ankona replied:
removed!

ashao committed:
Add integration tests to telemetry monitor

This PR adds 5 different integration tests that tests a variety of
common workflow patterns. These have not had the appropriate
test conditions set so these tests are currently passing
erroneously. Manual inspection of start/stop.json files shows that
only one entry is ever added.

ankona committed:
pass unique (non-final) step status_dir to indirect proxy

ankona committed:
log RC as integer to telemetry status updates

ankona committed:
fix missing var in test

ankona committed:
avoid repeated start tracking, addt'l details

ashao committed:
Add assertions to ensure models/db were launched correctly
Add ensemble test

MattToast committed:
rm dead code, fmt

ashao committed:
Correctly access single status returns from list
Add a colocated model example

MattToast committed:
fix tests so they use the new lmb api

MattToast committed:
controller api update to tests

MattToast committed:
Bring up to date with upstream

MattToast committed:
full WLM: make batch step meta look like normal step meta

ankona committed:
remove unused callbacks & associated tests

ankona committed:
add telemetry frequency to config

ankona committed:
avoid unmanaged tasks being monitored by job manager

ankona committed:
format telemetry outputs for readability

ankona committed:
enhance telemetry integration test assertions

ankona committed:
Merge branch 'oss517' into telem-dir-on-step

ankona committed:
try/cast tracked PIDs to int in unmanaged proxy

ankona committed:
add non-blocking test

ankona committed:
addt'l non-blocking test

ankona committed:
Merge pull request #3 from MattToast/telem-dir-on-step

Add Status Dir to Step Meta

ankona committed:
pass precomputed status_dir to telemetry proxy

ankona committed:
snooze nonblocking telemetry output test

ankona committed:
fix reversed step/task ID on managed tasks

ankona committed:
fix entity collection building bug

ankona committed:
add autoshutdown back to telmon

ankona committed:
adjust tests for modified deserialization code

ankona committed:
formatting

ankona committed:
addt'l tests

ankona committed:
skip unreliable tests

ankona committed:
linter

ankona committed:
linter

ankona committed:
remove shutdown tests

ankona committed:
avoid using old style namedtuple

ankona committed:
clean up import

ankona committed:
fix indirect test case, remove unused cli param

ankona committed:
use self ref to access cls method

ankona committed:
remove extraneous condition

ankona committed:
name private callback members appropriately

ankona committed:
move abstract method to correct base class

ankona committed:
remove deprecated arg

ankona committed:
add todo to avoid missing removal of method in bugfix

ankona committed:
remove unused constructor parameter

ankona committed:
remove commented code after refactor

ankona committed:
replace indirect.py looping for cleaner built-in

ankona committed:
use consistent value for indirect STEP_PID

ankona committed:
remove unused import

ankona committed:
fix docstring

ankona committed:
fix inline comment style

ankona committed:
replace print w/logging

ankona committed:
add long form args to indirect proxy

ankona committed:
remove short args for indirect proxy

ankona committed:
remove unnecessary condition

ankona committed:
replace clunky validation

ankona committed:
cleanup loop

ankona committed:
move telemetry env arg to config

ankona committed:
re-add check to avoid mypy err

ankona committed:
set generic type on popen

ankona committed:
remove unused ensembles key

ankona committed:
remove launcher type member var

ankona committed:
ensure set_launcher called on default launcher

ankona committed:
remove unused checks for non-JobEntity items

ankona committed:
fix attempt to use none retval, validate for mypy

ankona committed:
remove unused method & import

ankona committed:
remove unused internal type

ankona committed:
remove unused exp_path var

ankona committed:
use shared config instance

ankona committed:
remove confusing comment

ankona committed:
fix relative import

ankona committed:
start telmon in controller w/out jm callback

ankona committed:
typehint fix

ankona committed:
remove all unused job manager event hooks

ankona committed:
fix context manager bug, additional proxy logging

ankona committed:
typo

ankona committed:
remove deprecated method

ankona committed:
fix bad paths

ankona committed:
minor reversions

ankona committed:
use long arg names for telmon

ankona committed:
use  long params in calls to telmon entrypoint
